ngraph/core/reference/src/runtime/reference/non_max_suppression.cpp

```diff
@@ -0,0 +1,596 @@
+//*****************************************************************************
+// Copyright 2017-2020 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include "ngraph/op/non_max_suppression.hpp"
+#include <algorithm>
+#include <cmath>
+#include <queue>
+#include <vector>
+#include "ngraph/runtime/reference/non_max_suppression.hpp"
+#include "ngraph/shape.hpp"
+
+using namespace ngraph;
+using namespace ngraph::runtime::reference;
```

Comment:
General comment, could you add comments to the code?

Why do we need so much code and subsidiary structures?

I didn't review this file.

Reply:
This is an implementation according to ONNX NMS and TF NMSV5 implementations, and these implementations use subsidiary structure and such code. Next, to simplify calculations and to avoid templates, we need some preprocessing of input data of NMS-5, perform calculations for some precision (e.g., for `float`), and perform postprocessing of results of calculations.